### PR TITLE
Adding S4 Model

### PR DESCRIPTION
**Pull request by kamalkraj**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@Rocketknight1

### CODE REVIEW

**Comment by Rocketknight1** [comment]
@kamalkraj This looks great! Please let us know if you're encountering any problems, and of course ping me when it's ready for a final review.

**Review by sgugger — CHANGES_REQUESTED** [review]
Thanks for the PR, it's a complicated model you're adding!

The modeling file needs to be rewritten in depth before we can merge this PR IMO. This includes:
- using descriptive variable names: we can't use one-letter variable names with no comments in the code as we then won't be able to maintain code we do not understand
- use proper docstring that explain what the function/classes are doing
- don't use asserts but tests and raise proper errors of the right type.
- the building blocks of the model should all have names beginning with S4, like in other modeling files
- there shouldn't be code related to hyperparameters in the model, unless there is something very specific that requires it

File: docs/source/index.mdx [hunk 1]
```diff
@@ -247,6 +247,7 @@ Flax), PyTorch, and/or TensorFlow.
 |          RetriBERT          |       ✅       |       ✅       |       ✅        |         ❌         |      ❌      |
 |           RoBERTa           |       ✅       |       ✅       |       ✅        |         ✅         |      ✅      |
 |          RoFormer           |       ✅       |       ✅       |       ✅        |         ✅         |      ❌      |
+|             S4              |       ❌       |       ❌       |       ✅        |         ❌         |      ❌      |
```
[the remark this answers is not in the document]
Comment (kamalkraj) [reply docs/source/index.mdx:253]: done

File: docs/source/model_doc/s4.rst [hunk 2]
```diff
@@ -0,0 +1,68 @@
+..
```
[the remark this answers is not in the document]
Comment (kamalkraj) [reply docs/source/model_doc/s4.rst]: done

File: src/transformers/models/s4/configuration_s4.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+from ...configuration_utils import PretrainedConfig
+from ...utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+
+class S4Config(PretrainedConfig):
+    r"""
```
[the remark this answers is not in the document]
Comment (kamalkraj) [reply src/transformers/models/s4/configuration_s4.py:25]: Screenshot 2022-01-13 at 8 21 51 PM

ValueError: This script only converts rst files.

@sgugger

File: src/transformers/__init__.py [hunk 4]
```diff
@@ -1235,6 +1237,13 @@
             "load_tf_weights_in_roformer",
         ]
     )
+    _import_structure["models.s4"].extend(
+        [
+            "S4LMHeadModel",
+            "S4Model",
+            "S4PreTrainedModel",
+        ]
```
[the remark this answers is not in the document]
Comment (kamalkraj) [reply src/transformers/__init__.py:1251]: done

File: src/transformers/__init__.py [hunk 5]
```diff
@@ -2554,6 +2564,7 @@
         from .utils.dummy_pytorch_quantization_and_torch_objects import *
 
     if is_torch_available():
+
```
[the remark this answers is not in the document]
Comment (kamalkraj) [reply src/transformers/__init__.py]: done

File: src/transformers/__init__.py [hunk 6]
```diff
@@ -617,6 +618,7 @@
     _import_structure["modeling_utils"] = ["Conv1D", "PreTrainedModel", "apply_chunking_to_forward", "prune_layer"]
 
     # PyTorch models structure
+
```
[the remark this answers is not in the document]
Comment (kamalkraj) [reply src/transformers/__init__.py]: done.

File: src/transformers/models/s4/modeling_s4.py [hunk 7]
```diff
@@ -0,0 +1,2148 @@
...
+)
+from ...modeling_outputs import ModelOutput
+from ...modeling_utils import PreTrainedModel
+from ...utils import logging
+from .configuration_s4 import S4Config
+
+
+logger = logging.get_logger(__name__)
+
+# soft dependencies
```
[the remark this answers is not in the document]
Comment (kamalkraj) [reply src/transformers/models/s4/modeling_s4.py:46]: https://github.com/huggingface/transformers/blob/master/src/transformers/models/tapas/modeling_tapas.py#L52-L60

Should I change to OSError like above ?
@sgugger

[the remark this answers is not in the document]
Comment (kamalkraj) [reply src/transformers/models/s4/modeling_s4.py:46]: @Rocketknight1

**Comment by Rocketknight1** [comment]
Hi @kamalkraj ! We noticed there haven't been any commits here for a while - do you want us to find someone to take it over and resolve the issues from here? Either way, we appreciate everything you've done so far, so don't stress about it!

**Comment by kamalkraj** [comment]
Hi @Rocketknight1,

Got a little busy with some work. Will work on this week

**Comment by kamalkraj** [comment]
Sure @Rocketknight1. 
After the refractor also if you need any help from my side please ping here. 

Thanks

**Comment by gaceladri** [comment]
Hi, any news with this implementation? For some time I finished the @kamalkraj implementation, but I did not refactored the code. @Rocketknight1 was mentioning that the authors were planing to release a cleaner implementation. Do you have any update from the authors on it? also I think that it would be veeeeeery valuable if they also release his pre-trained model. Thanks!

**Comment by Rocketknight1** [comment]
@gaceladri This is something we're keeping an eye on. We were planning to do a port of the newer implementation, but we've seen [this paper ](https://twitter.com/ankgup2/status/1508807766093217804) recently. We believe this indicates that we're liable to see some significant updates to that architecture before the first foundation models arrive, and so we're backing away a little from porting S4 in its current form, at least until a significant foundation model arrives!

That said, we still think the architecture is likely to be very important, and we're seeing what we can do to get it to the point where those models exist. I know that's a little mysterious, but I don't want to over-promise anything just yet!

**Comment by gaceladri** [comment]
@Rocketknight1 Thanks a lot for the answers and for the papers as well. I had not seen it! Thanks!!

**Comment by bratao** [comment]
@Rocketknight1 do you think S4D is already mature enough to get into transformers? On https://github.com/HazyResearch/state-spaces/ it is already on version 3 and there is already a module with few dependencies https://github.com/HazyResearch/state-spaces/blob/main/src/models/s4/

I´m getting extraordinary results compared to regular transformers with small datasets

**Pull request closed without merging**